### ISSUE TITLE
generator-star-spacing: work better together with space-before-function-paren

### ISSUE THREAD

**Issue by jsnajdr** [eslint/eslint#6195]
**What version of ESLint are you using?**
2.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**What did you do? Please include the actual source code causing the issue.**
I'd like to define the following coding standard for functions and generators:

```
function name() {}
let f = function() {}
function* gen() {}
let g = function*() {}
let m = { * method() {} }
```

I.e.:
1. No space before function paren
2. One space after generator star, no space before
3. No space before function paren for anonymous generator

This is impossible to achieve with the current options for `space-before-function-parent` and `generator-star-spacing` rules. The closest I can get is this configuration:

```
{
  "parserOptions": { "ecmaVersion": 6 },
  "rules": {
    "space-before-function-paren": [ "error", "never" ],
    "generator-star-spacing": [ "error", "after" ]
  }
}
```

But this doesn't satisfy rule 3 above and forces me to use `function* ()`, which is inconsistent with `function()`.

The `generator-star-spacing` rule could use some `anonymous` vs. `named` distinction, similar to what `space-before-function-paren` already has.

**Comment by jryans** [comment]
(I am a member of the team @jsnajdr is representing above.)

I think for the fifth case of generator shorthands, we would actually prefer:

```
let m = { *method() {} }
```

so, the complete set would be:

```
function name() {}
let f = function() {}
function* gen() {}
let g = function*() {}
let m = { *method() {} }
```

To support this level of detail, I think `generator-star-spacing` would need to allow specifying star placement in 3 separate cases: `anonymous`, `named`, and `shorthand`. So, like @jsnajdr says, it could gain additional options similar to `space-before-function-paren`, but with an additional `shorthand` case as well.

**Comment by ilyavolodin** [comment]
@btmills @alberto is this something we want to add to the rule? (sorry, don't work with generators, so no idea).

**Comment by btmills** [comment]
> So, like @jsnajdr says, it could gain additional options similar to space-before-function-paren, but with an additional shorthand case as well.

Consistency with another rule sounds like a compelling argument to me. There are cases for normal functions that `space-before-function-parens` covers that we can't cover right now for generator functions because `generator-star-spacing` lacks similar options. I'm in favor 👍  @jsnajdr @jryans are either one of you willing to put together a PR for this if the issue is accepted?

**Comment by mysticatea** [comment]
:+1:

Also, please remember classes.
I would prefer `methods` instead of `shorthand`.

**Comment by jsnajdr** [comment]
@btmills Yes, I'm willing to create a PR that will add support for the following syntax options:
```
{"anonymous": "neither", "named": "after", "method": "neither"}
{"anonymous": {"before": false, "after": true}, ...}
```
The existing syntax will set the rules for all three forms, like it does now.
For `method`, the `before` space doesn't make sense, so it will be ignored.

@mysticatea Is there anything else we need to cover for classes? `method` option is enough, isn't it?

**Comment by mysticatea** [comment]
yes, method option is enough.

**Comment by jsnajdr** [comment]
There's one more case to consider: `class Foo { static *foo(){} }`

**Comment by mysticatea** [comment]
I think it's a method :)

**Comment by kaicataldo** [comment]
@jsnajdr Sounds like there was support for this initially. Are you still willing to work on this?

**Comment by iTob191** [comment]
For `let g = function*() {}`:
Maybe it would be sufficient if `space-before-function-paren` would take precedence over `generator-star-spacing`. I can't imagine that anyone wants to use `function()` and `function* ()`.

**Comment by kaicataldo** [comment]
I'm closing this issues because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
